File: syslog-ng/modules/native/parser.c
/*
 * Copyright (c) 2015 BalaBit
 * Copyright (c) 2015 Tibor Benke
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include "parser.h"
#include <assert.h>

struct NativeParserProxy;

__attribute__((visibility("hidden"))) void
native_parser_proxy_free(struct NativeParserProxy *this);

__attribute__((visibility("hidden"))) gboolean
native_parser_proxy_set_option(struct NativeParserProxy *self, const gchar *key, const gchar *value);

__attribute__((visibility("hidden"))) gboolean
native_parser_proxy_process(struct NativeParserProxy *this, LogParser *super, LogMessage *pmsg, const gchar *input);

__attribute__((visibility("hidden"))) int
native_parser_proxy_init(struct NativeParserProxy *s);

__attribute__((visibility("hidden"))) int
native_parser_proxy_deinit(struct NativeParserProxy *s);

__attribute__((visibility("hidden"))) struct NativeParserProxy *
native_parser_proxy_new(GlobalConfig *cfg);

__attribute__((visibility("hidden"))) struct NativeParserProxy *
native_parser_proxy_clone(struct NativeParserProxy *self);

typedef struct _ParserNative
{
  LogParser super;
  struct NativeParserProxy *native_object;
} ParserNative;

static LogPipe *
native_parser_clone(LogPipe *s);

static gboolean
native_parser_process(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_options, const gchar *input,
                      gsize input_len)
{
  ParserNative *self = (ParserNative *) s;

  LogMessage *writable_msg = log_msg_make_writable(pmsg, path_options);
  return native_parser_proxy_process(self->native_object, &self->super, writable_msg, input);
}

gboolean
native_parser_set_option(LogParser *s, gchar *key, gchar *value)
{
  ParserNative *self = (ParserNative *)s;

  return native_parser_proxy_set_option(self->native_object, key, value);
}

static gboolean
native_parser_init(LogPipe *s)
{
  ParserNative *self = (ParserNative *) s;

  return log_parser_init_method(s) && native_parser_proxy_init(self->native_object);
}

static gboolean
native_parser_deinit(LogPipe *s)
{
  ParserNative *self = (ParserNative *) s;

  if (!native_parser_proxy_deinit(self->native_object))
    return FALSE;
  return log_parser_deinit_method(s);
}

static void
native_parser_free(LogPipe *s)
{
  ParserNative *self = (ParserNative *)s;

  native_parser_proxy_free(self->native_object);
  log_parser_free_method(s);
}

static void
__setup_callback_methods(ParserNative *self)
{
  self->super.process = native_parser_process;
  self->super.super.free_fn = native_parser_free;
  self->super.super.clone = native_parser_clone;
  self->super.super.init = native_parser_init;
  self->super.super.deinit = native_parser_deinit;
}

static LogPipe *
native_parser_clone(LogPipe *s)
{
  ParserNative *self = (ParserNative *) s;
  ParserNative *cloned;

  cloned = (ParserNative *) g_new0(ParserNative, 1);
  log_parser_init_instance(&cloned->super, s->cfg);
  cloned->native_object = native_parser_proxy_clone(self->native_object);
  assert(self != cloned);

  if (!cloned->native_object)
    {
      g_free(cloned);
      return NULL;
    }

  __setup_callback_methods(cloned);

  return &cloned->super.super;
}

__attribute__((visibility("hidden"))) LogParser *
native_parser_new(GlobalConfig *cfg)
{
  ParserNative *self = (ParserNative *) g_new0(ParserNative, 1);

  log_parser_init_instance(&self->super, cfg);
  self->native_object = native_parser_proxy_new(cfg);

  if (!self->native_object)
    {
      g_free(self);
      return NULL;
    }

  __setup_callback_methods(self);

  return &self->super;
}

File: modules/python-modules/syslogng/parser.py
#############################################################################
# Copyright (c) 2022 Balazs Scheidler <bazsi77@gmail.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
# As an additional exemption you are allowed to compile & link against the
# OpenSSL libraries as published by the OpenSSL project. See the file
# COPYING for details.
#
#############################################################################
# pylint: disable=function-redefined

try:
    from _syslogng import LogParser

except ImportError:
    import warnings
    warnings.warn("You have imported the syslogng package outside of syslog-ng, "
                  "thus some of the functionality is not available. "
                  "Defining fake classes for those exported by the underlying syslog-ng code")

    LogParser = object


class LogParser(LogParser):
    def init(self, options):
        """This method is called when the configuration is initialized

        Args:
            options:
                is a dictionary that contains options passed using the
                options() parameter of the python() destination

        Returns:
            bool: True if successful, False otherwise
        """
        return True

    def deinit(self):
        """This method is called when the configuration is deinitialized"""

    def parse(self, msg):
        """This method is called to 'parse' the message

        Args:
            msg: LogMessage
                is a syslog-ng message that allows dict-like access to
                name-value pairs.

        Returns:
            bool: True if parsing was successful, False to drop the message
        """
        raise NotImplementedError
